tests/test_PR045_flatten_operation.py

```diff
@@ -0,0 +1,127 @@
+            if all(isinstance(x, list) for x in data):
+                return sum(data, [])
+            else:
+                raise ValueError("cannot concatenate non-lists")
+        else:
+            return [flatten(x, axis - 1) for x in data]
+    else:
+        raise ValueError("cannot flatten {0} objects".format(type(data)))
+
+def test_flatten_empty_array():
+    empty = awkward1.layout.EmptyArray()
+
+    assert awkward1.tolist(empty) == []
+    assert awkward1.tolist(empty.flatten()) == []
+
+def test_flatten_list_array():
+    content = awkward1.layout.NumpyArray(numpy.array([0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]))
+    starts  = awkward1.layout.Index64(numpy.array([0, 3, 3, 5, 6]))
+    stops   = awkward1.layout.Index64(numpy.array([3, 3, 5, 6, 9]))
+    array   = awkward1.layout.ListArray64(starts, stops, content)
+
+    assert awkward1.tolist(array) == [[0.0, 1.1, 2.2], [], [3.3, 4.4], [5.5], [6.6, 7.7, 8.8]]
+    assert awkward1.tolist(array.flatten()) == [0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8]
+
+    array2 = array[2:-1]
+    assert awkward1.tolist(array2.flatten()) == [3.3, 4.4, 5.5]
+
+    # The following are allowed:
+    #     * out of order (4:7 before 0:1)
+    #     * overlaps (0:1 and 0:4 and 1:5)
+    #     * beyond content starts[i] == stops[i] (999)
+    #
+    # See https://github.com/scikit-hep/awkward-1.0/wiki/ListArray.md
+    #
+    content = awkward1.layout.NumpyArray(numpy.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=numpy.int64))
+    starts  = awkward1.layout.Index64(numpy.array([4, 999, 0, 0, 1, 7]))
+    stops   = awkward1.layout.Index64(numpy.array([7, 999, 1, 4, 5, 10]))
+    array   = awkward1.layout.ListArray64(starts, stops, content)
+    assert awkward1.tolist(array) == [[4, 5, 6], [], [0], [0, 1, 2, 3], [1, 2, 3, 4], [7, 8, 9]]
+    assert flatten(awkward1.tolist(array)) == [4, 5, 6, 0, 0, 1, 2, 3, 1, 2, 3, 4, 7, 8, 9]
+    # Flatten should handle this.
+    # assert awkward1.tolist(array.flatten()) == [4, 5, 6, 0, 0, 1, 2, 3, 1, 2, 3, 4, 7, 8, 9]
+
+def test_flatten_list_offset_array():
+    content = awkward1.layout.NumpyArray(numpy.array([0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]))
+    offsets = awkward1.layout.Index64(numpy.array([0, 3, 3, 5, 6, 10], dtype=numpy.int64))
+    array   = awkward1.layout.ListOffsetArray64(offsets, content)
+
+    assert awkward1.tolist(array) == [[0.0, 1.1, 2.2], [], [3.3, 4.4], [5.5], [6.6, 7.7, 8.8, 9.9]]
+    assert awkward1.tolist(array.flatten()) == [0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]
+
+    array2 = array[2:-1]
+    assert awkward1.tolist(array2.flatten()) == [3.3, 4.4, 5.5]
+
+def test_flatten_numpy_array():
+    dint64 = awkward1.layout.PrimitiveType("int64", {"__class__": "D", "__str__": "D[int64]"})
+    array = awkward1.layout.NumpyArray(numpy.arange(2*3*5, dtype=numpy.int64).reshape(2, 3, 5)).astype(awkward1.layout.RegularType(awkward1.layout.RegularType(dint64, 5), 3))
+    print(awkward1.tolist(array))
```

Comment:
Creating type parameters and adding them is not wrong, but also not necessary. The following (simpler) setup is equivalent for these purposes:

```python
array = awkward1.layout.NumpyArray(numpy.arange(2*3*5, dtype=numpy.int64).reshape(2, 3, 5))
```